libsrc/c64/ser/c64-up2400.s

```diff
@@ -0,0 +1,498 @@
+
+; Jump table
+
+        .word   SER_INSTALL
+        .word   SER_UNINSTALL
+        .word   SER_OPEN
+        .word   SER_CLOSE
+        .word   SER_GET
+        .word   SER_PUT
+        .word   SER_STATUS
+        .word   SER_IOCTL
+        .word   SER_IRQ
+
+.bss
+
+.data
+
+; FIXME: these are likely calculated for NTSC?
+strt24:  .word $01cb     ; 459   start bit times
+strt12:  .word $0442     ; 1090
+strt03:  .word $1333     ; 4915
+
+full24:  .word $01a5     ; 421   full bit times
+full12:  .word $034d     ; 845
+full03:  .word $0d52     ; 3410
+
+.rodata
+
+baudrate = 10
+databits = 0
+stopbit = 0
+
+wire = 1
+duplex = 0
+parity = 0
+
+serial_config:
+        .byte baudrate + databits + stopbit
+        .byte wire + duplex + parity
+
+.code
+
+;----------------------------------------------------------------------------
+; INSTALL routine. Is called after the driver is loaded into memory. If
+; possible, check if the hardware is present.
+; Must return an SER_ERR_xx code in a/x.
+
+SER_INSTALL:
+        lda     #<SER_ERR_OK
+        tax
+        rts
+
+;----------------------------------------------------------------------------
+; UNINSTALL routine. Is called before the driver is removed from memory.
+; Must return an SER_ERR_xx code in a/x.
+
+SER_UNINSTALL:
+        lda     #<SER_ERR_OK
+        tax
+        rts
+
+;----------------------------------------------------------------------------
+; PARAMS routine. A pointer to a ser_params structure is passed in ptr1.
+; Must return an SER_ERR_xx code in a/x.
+
+SER_OPEN:
+        lda #2
+        ldx #<serial_config
+        ldy #>serial_config
+        jsr SETNAM
+
+        lda #2
+        tax
+        tay
+        jsr SETLFS
+        jsr OPEN
+
+        jsr setup
+
+        lda #<SER_ERR_OK
+        tax
+        rts
+
+;----------------------------------------------------------------------------
+; CLOSE: Close the port, disable interrupts and flush the buffer. Called
+; without parameters. Must return an error code in a/x.
+;
+; We can omit close now, save for cleaning up ;-)
+SER_CLOSE:
+        lda     #<SER_ERR_OK
+        tax
+        rts
+
+;----------------------------------------------------------------------------
+; GET: Will fetch a character from the receive buffer and store it into the
+; variable pointer to by ptr1. If no data is available, SER_ERR_NO_DATA is
+; return.
+;
+
+SER_GET:
+        ldx #2
+        jsr CHKIN
+        jsr rshavedata
+        beq SER_GET_NO_DATA
+        jsr SER_OURBASIN
+        ldx #$00
+        sta (ptr1,x)
+        jsr CLRCH
+        lda #<SER_ERR_OK
+        tax
+        rts
+
+SER_GET_NO_DATA:
+        lda #<SER_ERR_NO_DATA
+        ldx #>SER_ERR_NO_DATA
+        rts
+
+;----------------------------------------------------------------------------
+; OURBASIN: This is a minimised call to get the character from the buffer.
+; The Kernal code does not allow zero bytes (0x00)... this does.
+;
+
+SER_OURBASIN:
+        jsr $F14E
```

Comment:
Calling directly into ROM?
I think at least a comment is missing what these function are supposed to do (also below $F1B4).
Will this work on e.g. VIC20?